crates/hir/src/source_analyzer.rs

```diff
@@ -807,7 +818,7 @@ impl SourceAnalyzer {
                     return Some((PathResolution::Def(AssocItem::from(assoc).into()), Some(subst)));
                 }
                 if let Some(VariantId::EnumVariantId(variant)) =
-                    infer.variant_resolution_for_pat(pat_id)
+                    infer.variant_resolution_for_pat(pat_id.as_pat()?)
```

Comment:
Similar thing here, might need to handle the expression case due to having a path pattern